nm: Introduce new option to enable trace log on second try

cathay4t opened the pull request:
Introducing new API:
 * Rust: `NetworkState::set_verbose_log_when_retry(bool)`
 * C: `NMSTATE_FLAG_VERBOSE_WHEN_RETRY`
 * Python: `verbose_on_retry` argument of `libnmstate.apply()`.

When enabled, before retrying the network state apply, enable
trace log of NetworkManager and restore it after finished.

By default, this action is disabled.

The trace log of NetworkManager could be very helpful for debugging failure.

Integration test case included.

ihuguet commented:
I don't think this is a good idea. TRACE is *very* verbose and it can even affect the behavior of NM making it noticeable slower. Also, DEBUG and TRACE logs are only useful for developers. On the other side, they can be confusing for users. Also, users may get flooded by log messages, so they cannot identify which are the relevant ones.

I agree that this can be very useful for us when we have to investigate something, bug it should be optionally enabled somehow, IMO.

cathay4t commented:
The trace log only enabled on second retry which often means something wrong happens. And nmstate will restore the log setting after apply.

For performance concern, I think it is acceptable for user to have performance penalty when a network change error is happening.

Yes, I will change it to be a option/flag and suggest kubernets-nmstate to enable it.

cathay4t committed:
nm: Introduce new option to enable trace log on second try

Introducing new API:
 * Rust: `NetworkState::set_verbose_log_when_retry(bool)`
 * C: `NMSTATE_FLAG_VERBOSE_WHEN_RETRY`
 * Python: `verbose_on_retry` argument of `libnmstate.apply()`.

When enabled, before retrying the network state apply, enable
trace log of NetworkManager and restore it after finished.

By default, this action is disabled.

The trace log of NetworkManager could be very helpful for debugging failure.

Integration test case included.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
New patch uploaded to change the action as option and disabled by default.

ihuguet commented on rust/src/lib/net_state.rs:
```diff
@@ -325,21 +322,20 @@ pub(crate) struct MergedNetworkState {
     pub(crate) routes: MergedRoutes,
     pub(crate) rules: MergedRouteRules,
-    pub(crate) memory_only: bool,
+    pub(crate) option: NetworkStateOption,
```
If only `memory_only` is applicable to `MergedNetworkState`, we shouldn't include the whole `NetworkStateOption`. It will become confusing to know where you have to set or read `no_verify` and others. From `NetworkState` or from `MergedNetworkState`?

(Actually, now that I look at it, it is a bit weird that they are in `NetworkState` too. They are not part of the network state. They are options about how to apply a new network state and they make sense for a **desired** network state, but what do they mean for a **current** network state? I'm just thinking about it, I'm not saying that we should change it now).

ihuguet commented:
Another thing I'm thinking is that, for debugging purposes, although this is better than before, probably the more relevant log messages will happen on the first try. Is there any way for kubernetes-nmstate users to enable trace logging since the beginning if we request them trace logs?

cathay4t commented:
> Another thing I'm thinking is that, for debugging purposes, although this is better than before, probably the more relevant log messages will happen on the first try. Is there any way for kubernetes-nmstate users to enable trace logging since the beginning if we request them trace logs?

This is forcing 90% normal user to pay the penalty of performance degrade on faults they will never got/face.

I will not recommend k8s-nmstate to enable trace log when nmstatectl is applying. Please be aware k8s-nmstate apply NNCP more often than we expected(label change, pod restart, node restart, and etc), so enable trace at the beginning of nmstatectl process will cause performance issue to production workloads.

ihuguet commented:
What I say if there is a way to enable them **if we request trace logs**. Not always, only for certain cases.

cathay4t commented:
> What I say if there is a way to enable them **if we request trace logs**. Not always, only for certain cases.

We need new PR or issue to discuss the domain of `certain cases`.

cathay4t committed:
Merge branch 'base' into enable_nm_log_when_failure

cathay4t commented:
I lose interest on this and worry about degrading performance of user's OS. Closing.
